### PR TITLE
concretizer: convert virtuals to facts and move all rules to `concretize.lp`

### PR DESCRIPTION
**Pull request by tgamblin**
This converts the virtual handling in the new concretizer from already-ground rules to facts. This is the last thing that needs to be refactored, and it converts the entire concretizer to just use facts.

The previous way of handling virtuals hinged on rules involving `single_provider_for` facts that were tied to the virtual and a version range. The new method uses the condition pattern we've been using for dependencies, externals, and conflicts.

To handle virtuals as conditions, we impose constraints on "fake" virtual specs in the logic program. i.e., `version_satisfies("mpi", "2.0:", "2.0")` is legal whereas before we wouldn't have seen something like
this. Currently, constriants are only handled on versions -- we don't handle variants or anything else yet, but they key change here is that we *could*. For a long time, virtual handling in Spack has only dealt with versions, and we'd like to be able to handle variants as well. We could easily add an integrity constraint to handle variants like the one we use for versions.

This is what the new logic looks like:

```prolog
dependency_conditions_hold(ID, Package, Virtual) :-
  attr(Name, Arg1)             : required_provider_condition(ID, Name, Arg1);
  attr(Name, Arg1, Arg2)       : required_provider_condition(ID, Name, Arg1, Arg2);
  attr(Name, Arg1, Arg2, Arg3) : required_provider_condition(ID, Name, Arg1, Arg2, Arg3);
  virtual_node(Virtual);
  provider_condition(ID, Package, Virtual).
```

What this does is a bit confusing -- the conditions are evaluated on the provider, and they're *imposed* like dependency conditions.  Provider conditions declare `imposed_dependency_condition` like conditional dependencies, and they rely on the same rules to enforce the imposed constraints:

```prolog
attr(Name, Arg1, Arg2) :-
  dependency_conditions_hold(ID, Package, Dependency),
  imposed_dependency_condition(ID, Name, Arg1, Arg2).

attr(Name, Arg1, Arg2, Arg3) :-
  dependency_conditions_hold(ID, Package, Dependency),
  imposed_dependency_condition(ID, Name, Arg1, Arg2, Arg3).
```

### CODE REVIEW

File: lib/spack/spack/solver/asp.py [hunk 1]
```diff
@@ -37,11 +37,7 @@
 import spack.package_prefs
 import spack.repo
 import spack.variant
-from spack.version import ver
-
-
-#: max line length for ASP programs in characters
-_max_line = 80
+import spack.version
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:40]: I know it's minor, but I appreciate removing `from` import 🙂

File: lib/spack/spack/solver/asp.py [hunk 2]
```diff
@@ -698,19 +629,74 @@ def pkg_rules(self, pkg, tests):
             )
         )
 
+    def _condition_facts(
+            self, pkg_name, cond_spec, dep_spec,
+            cond_fn, require_fn, impose_fn
+    ):
+        """Generate facts for a dependnecy or virtual provider condition.
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py]:
```suggestion
        """Generate facts for a dependency or virtual provider condition.
```

File: lib/spack/spack/solver/asp.py [hunk 3]
```diff
@@ -698,19 +629,74 @@ def pkg_rules(self, pkg, tests):
...
+
+        Arguments:
+            pkg_name (str): name of the package that triggers the
+                condition (e.g., the dependent or the provider)
+            cond_spec (Spec): the dependency spec representing the
+                condition that needs to be True (can be anonymous)
+            dep_spec (Spec): the sepc of the dependnecy or provider
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py]:
```suggestion
            dep_spec (Spec): the spec of the dependency or provider
```

File: lib/spack/spack/solver/asp.py [hunk 4]
```diff
@@ -1232,26 +1174,39 @@ def define_version_constraints(self):
...
+                result = [v.start] if v.start else []
+                result += [v.end] if v.end else []
+                return result
+            elif isinstance(v, spack.version.VersionList):
+                return sum((versions_for(e) for e in v), [])
+            else:
+                raise TypeError("expected version type, found: %s" % type(v))
+
+        # define a set of synthetic possible versions for virtuals, so
+        # that `version_satisfies(Pakcage, Constraint, Version)` has the
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py]:
```suggestion
        # that `version_satisfies(Package, Constraint, Version)` has the
```

**Review by alalazo — COMMENTED** [review]
LGTM. Only caught a few typos.

**Comment by tgamblin** [comment]
@alalazo fixed

**Review by alalazo — APPROVED** [review]

**Pull request merged**